lib/models.py

```diff
@@ -2,13 +2,267 @@
+                at this level, the language model ranks them in order
+                of likelihood of answering the problem correctly
+            (3) the algorithm selects a set of thoughts to keep based 
+                on the above scoring
+
+        Returns a list of thoughts, ys, which are the most likely 
+        candidates to being correct chains of thoughts to solve the 
+        problem at the bottom of the tree.
+        """
+
+        ys = [''] #current output candidates
+        infos = []
+
+        for step in range(self.num_steps):
+
+            #generation
+            new_ys = [self.get_samples(x, y) for y in ys]
+            new_ys = list(itertools.chain(*new_ys))
+            ids = list(range(len(new_ys)))
+
+            #evaluation 
+            values = [self.get_score(x, y) for y in new_ys]
+
+            #selection
+            if self.selection_method == 'greedy':     
+                # gets the best N=num_select_sample choices
+                select_ids = sorted(ids, key=lambda x: values[x], reverse=True)[:self.num_select_sample]
+            elif self.selection_method == 'sample':
+                # weighted samples from all possible options to get N=num_select_sample choices.
+                ps = np.array(values) / sum(values)
+                select_ids = np.random.choice(ids, size=self.num_select_sample, p=ps).tolist()
+            select_new_ys = [new_ys[select_id] for select_id in select_ids]
+            select_new_values = [values[select_id] for select_id in select_ids]
+
+            #logging
+            infos.append({'step':step, 'x':x, 'ys':ys, 'new_ys':new_ys, 'values':values, 'select_new_ys':select_new_ys})
+
+            #preparation for next loop
+            ys = select_new_ys
+        
+        if self.verbose:
+            print("INFO: Chain of thought information ", infos)
+
+        return ys, select_new_values
+    
+    def __call__(self, prompt):
+        """
+        Calls the Chain of thoughts algorithm to get list of most
+        likely thoughts.
+
+        After, the method asks a final answer evaluator to get the 
+        final answer from the best chainof thoughts produced at the 
+        lowest level of the tree. 
+
+        Returns a string containing the answer to the prompt
+        """
+        ys, values = self.solve(prompt)
+        ids = list(range(len(ys)))
+
+        #selection to get best final answer
+        if self.selection_method == 'greedy':     
+            # gets the best N=num_select_sample choices
+            select_ids = sorted(ids, key=lambda x: values[x], reverse=True)[:1]
+        elif self.selection_method == 'sample':
+            # weighted samples from all possible options to get N=num_select_sample choices.
+            ps = np.array(values) / sum(values)
+            select_ids = np.random.choice(ids, size=1, p=ps).tolist()
+        final_CoT = [ys[select_id] for select_id in select_ids][0]
```

Comment:
Why not `final_CoT = ys[select_ids[0]]`?

Reply:
Good idea! Thanks.